rfcs/0000-rfc-calling-context-attribute.md

```diff
@@ -0,0 +1,116 @@
+- Feature Name: calling-context-attribute
+- Start Date: 2022-07-06
+- RFC PR: [FuelLabs/sway-rfcs#0000](https://github.com/FuelLabs/sway-rfcs/pull/9)
+- Sway Issue: [FueLabs/sway#963](https://github.com/FuelLabs/sway/issues/963)
+
+# Summary
+
+[summary]: #summary
+
+Sway will support an attribute (i.e. `#[calling_context(contract_only)]`) to annotate the supported calling contexts for particular functions.
+
+This annotation can then be used by the compiler to check and restrict the calling of these functions to a compatible calling context.
+
+# Motivation
+
+[motivation]: #motivation
+
+The main goal is to be able to get compile-time safety when working with functions that should only be called from a contract context.
+
+A real-world example of this is the [msg_amount() stdlib function](https://github.com/FuelLabs/sway-lib-std/blob/6a8f5bf588df5d0679fb834f05c900f2e54de426/src/context.sw#L33-L38) which only returns a proper value when used in a contract context. In an external context is returns a well-defined but not semantically representative value, specifically 0, as explained by @adlerjohn in [FueLabs/sway#963](https://github.com/FuelLabs/sway/issues/963).
```

Comment:
I'm curious to see some more real-world examples of cases where we'd like to use this attribute.

If cases like using `msg_amount` are the only kinds of cases we have in mind, perhaps as an alternative we should consider having users take `msg_amount` via a function argument instead?

E.g. rather than

```rust
#[calling_context(contract_only)]
fn bar() {
    let msg_amt = msg_amount();
    // ...
}
```

Library writers can do:

```rust
fn bar(msg_amt: u64) {
    // ...
}
```

Of course, in contract code, the caller would then need to call `bar` with the explicit argument `bar(msg_amount())` rather than just `bar()`, however the former is arguably better for the reader as they can see where the value is coming from, rather than having `bar` introduce the side-effect?

I'm weary that our current motivations might not fully justify introducing a whole new attribute, but perhaps a more motivating example might help?

Reply:
Whoops, looks like this conversation was never resolved.

If we always defer the acquisition of the side-effectual information via arguments, we still eventually have to write functions in the standard library which produce this information and we want to gate access to them at compile time. This _could_ be accomplished with conditional compilation, but the error messages for that would not be great. We could improve them with conditional compilation _and_ a `compile_error!()` macro like Rust has?

E.g.

```rust
// in stdlib
#[cfg(not(external))]
fn msg_amount() {
  compile_error!("Message amount cannot be called from a non-external context.");
}

```

Reply:
I can see us eventually wanting conditional compilation for sure, however this issue in particular (and potentially this entire RFC) might be better solved by https://github.com/FuelLabs/sway/issues/2463.

That is, if the contract-specific functions were moved to contract methods, it would be impossible to call them in an incorrect context. This way we have the added benefit that our contract-only functions also are not side-effecting, and it's always clear when a function might behave differently as a result of referencing some contract or msg state, as either the contract or the data itself would need to be passed as an argument.